There are 20531 features (genes) with only 640 training samples. As there are far more variables than samples the model can fit noise in the training data and fail on new patients due to the the curse of dimensionality. Feature selection is required to reduce the number of features.

The two main stages of feature selection will be done by:
1) Unsupervised variance filter - removing the features which have no variance across all of the samples having no differentiation in value between the different genes providing the model no value.
2) 

In [1]:
pip install -r requirements.txt

Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 25.1.1 -> 26.1.2
[notice] To update, run: python.exe -m pip install --upgrade pip


In [2]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

In [3]:
x_train = pd.read_csv('../data/processed/x_train.csv', index_col=0)
y_train = pd.read_csv('../data/processed/y_train.csv')['label'].values
x_test = pd.read_csv('../data/processed/x_test.csv', index_col=0)
y_test = pd.read_csv('../data/processed/y_test.csv')['label'].values

import joblib
le = joblib.load('../data/processed/label_encoder.joblib')
print(f"X_train shape: {x_train.shape}")
print(f"X_test shape:  {x_test.shape}")
print(f"y_train shape: {y_train.shape}")
print(f"y_test shape:  {y_test.shape}")
print(f"Classes: {le.classes_}")

X_train shape: (640, 20531)
X_test shape:  (161, 20531)
y_train shape: (640,)
y_test shape:  (161,)
Classes: ['BRCA' 'COAD' 'KIRC' 'LUAD' 'PRAD']


In [4]:
var_scaled = x_train.var()
print(var_scaled.describe())

count    20531.000000
mean         0.988101
std          0.115345
min          0.000000
25%          1.001565
50%          1.001565
75%          1.001565
max          1.001565
dtype: float64


Every quartile is equal since we had normalised the gene variance to variance ~1 across the samples (not exactly 1 as explained earlier). With genes at variance 0 that are completely dead across all the samples.

In [5]:
print(f'genes with no variance: {(var_scaled == 0).sum()}')

genes with no variance: 276


There are 276 dead genes with no variance ready to be removed.

In [6]:
var_scaled = x_train.var()
kept_genes = x_train.columns[var_scaled > 0]

x_train = x_train[kept_genes]
x_test = x_test[kept_genes]

print(len(kept_genes))

20255
